src/main/java/software/amazon/smithy/lsp/ext/SmithyProject.java

```diff
@@ -0,0 +1,209 @@
+        List<Path> imports = config.getImports().stream().map(p -> Paths.get(root.getAbsolutePath(), p).normalize())
+                .collect(Collectors.toList());
+
+        if (!imports.contains(Paths.get("."))) {
+            imports.add(root.toPath());
+        }
+
+        LspLog.println("Imports from config: " + imports + " will be resolved against root " + root);
+
+        List<File> smithyFiles = discoverSmithyFiles(imports, root);
+        LspLog.println("Discovered smithy files: " + smithyFiles);
+
+        List<File> externalJars = downloadExternalDependencies(config);
+        LspLog.println("Downloaded external jars: " + externalJars);
+
+        return load(imports, smithyFiles, externalJars, root);
+
+    }
+
+    private static Either<Exception, SmithyProject> load(List<Path> imports, List<File> smithyFiles,
+                                                         List<File> externalJars, File root) {
+        Either<Exception, ValidatedResult<Model>> model = createModel(smithyFiles, externalJars);
+
+        if (model.isLeft()) {
+            return Either.forLeft(model.getLeft());
+        } else {
+            return Either.forRight(new SmithyProject(imports, smithyFiles, externalJars, root, model.getRight()));
+        }
+    }
+
+    private static Either<Exception, ValidatedResult<Model>> createModel(List<File> discoveredFiles,
+                                                                         List<File> externalJars) {
+        return SmithyInterface.readModel(discoveredFiles, externalJars);
+    }
+
+    private static Map<String, List<Location>> collectLocations(Model model) {
+        Map<String, List<Location>> locations = new HashMap<>();
+        model.shapes().forEach(shape -> {
+            SourceLocation sourceLocation = shape.getSourceLocation();
+            String uri = sourceLocation.getFilename();
+            if (uri.startsWith("jar:file:")) {
+                uri = "smithyjar:" + uri.substring(9);
+            } else if (!uri.startsWith("file:")) {
+                uri = "file:" + uri;
+            }
+            Position pos = new Position(sourceLocation.getLine() - 1, sourceLocation.getColumn() - 1);
+            Location location = new Location(uri, new Range(pos, pos));
+
+            String shapeName = shape.getId().getName();
+            // Members get the same shapeName as their parent structure
+            // so we ignore them, to avoil producing a location per-member
+            // TODO: index members somehow as well?
+            if (shape.getType() != ShapeType.MEMBER) {
+                if (locations.containsKey(shapeName)) {
+                    locations.get(shapeName).add(location);
+                } else {
+                    List<Location> locList = new ArrayList<Location>();
+                    locList.add(location);
+                    locations.put(shapeName, locList);
+                }
+            }
```

Comment:
This is a fixed version of location tracker, which doesn't have weird location jumps and correctly tracks things named the same, but in different namespaces.

Reply:
Happy with that for the time being. In theory we could could push things really far and index the location of the traits, as each node (json ADT) contains [location information](https://github.com/awslabs/smithy/blob/main/smithy-model/src/main/java/software/amazon/smithy/model/node/Node.java#L49), which is probably the reason behind smithy having its own json library.